### PR TITLE
X509 Implementation and tests

### PR DESCRIPTION
**Pull request by anjuthomas**

### CODE REVIEW

File: src/controllers/messageTypes/jwt/jsonwebtoken/decode.js [hunk 1]
```diff
@@ -9,8 +9,6 @@ var jwtVerifier       = require('./verify')
 var jwkToPem          = require('jwk-to-pem');
 var forge = require('node-forge');
 
-
-
 var jwtDecoder = JWTDecoder.prototype;
```
Comment (lyzhang27) [inline src/controllers/messageTypes/jwt/jsonwebtoken/decode.js]: Should we remove this and just define functions with "JWTDecoder.prototype"?

File: src/controllers/messageTypes/jwt/jsonwebtoken/decode.js [hunk 2]
```diff
@@ -118,14 +116,34 @@ jwtDecoder.parsePayload = function(payload){
   }
 }
 
+jwtDecoder.verifyJwtSign = function(token, secretOrPublicKey, tokenProfile, otherOptions, baseEncoding, callback){
```
Comment (lyzhang27) [inline src/controllers/messageTypes/jwt/jsonwebtoken/decode.js]: JWTDecoder.prototype?

File: src/controllers/messageTypes/jwt/jsonwebtoken/decode.js [hunk 3]
```diff
@@ -118,14 +116,34 @@ jwtDecoder.parsePayload = function(payload){
...
+      }
+            
+      if (!valid){
+        console.log('invalid signature');
+        return false;
+      } else {
+        return true;
+      }
+    }
+}
```
Comment (lyzhang27) [inline src/controllers/messageTypes/jwt/jsonwebtoken/decode.js]: Indentation is off here

File: src/controllers/messageTypes/jwt/jsonwebtoken/decode.js [hunk 4]
```diff
@@ -9,8 +9,6 @@ var jwtVerifier       = require('./verify')
 var jwkToPem          = require('jwk-to-pem');
```
Comment (lyzhang27) [inline src/controllers/messageTypes/jwt/jsonwebtoken/decode.js]: Missing ";" in line 8

File: src/controllers/messageTypes/jwt/jsonwebtoken/decode.js [hunk 5]
```diff
@@ -118,14 +116,34 @@ jwtDecoder.parsePayload = function(payload){
...
+      return false;
+    } else {
+      var valid = null;
+      try {
+        valid = jws.verify(token, jwt.header.alg, secretOrPublicKey, baseEncoding);
+      } catch (e) {
+        return done(e);
+      }
+            
+      if (!valid){
```
Comment (lyzhang27) [inline src/controllers/messageTypes/jwt/jsonwebtoken/decode.js]: Missing space after the ")"

File: src/controllers/messageTypes/jwt/jsonwebtoken/decode.js [hunk 6]
```diff
@@ -137,20 +155,31 @@ jwtDecoder.verifyJwtSignature = function(token, secretOrPublicKey, tokenProfile,
...
               // no matching kid 
+              return false;
             }
           } else {
             // try using the first key
             const first = JSON.parse(certs).keys[0];
             var pemFromX509 = this.convertX509ToPem(first);
             console.log(pemFromX509);
-            return jws.verify(token, jwt.header.alg, pemFromX509, baseEncoding);
+            //return jws.verify(token, jwt.header.alg, secretOrPublicKey, baseEncoding);
```
Comment (lyzhang27) [inline src/controllers/messageTypes/jwt/jsonwebtoken/decode.js]: Why are these two lines commented out?

File: src/controllers/messageTypes/jwt/jsonwebtoken/decode.js [hunk 7]
```diff
@@ -137,20 +155,31 @@ jwtDecoder.verifyJwtSignature = function(token, secretOrPublicKey, tokenProfile,
...
             }
+            
+            if (!valid){
+              console.log('invalid signature');
+              return false;
+            } else {
+              return true;
+            }
+          }
         }
```
Comment (lyzhang27) [inline src/controllers/messageTypes/jwt/jsonwebtoken/decode.js]: Indentation is off here.

File: test/x509-test.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
+var fs = require('fs');
+var BasicIdToken = require('../src/models/tokenProfiles/basicIdToken');
+var assert = require('chai').assert;  
+var XMLHttpRequest = require("xmlhttprequest").XMLHttpRequest;
+var jwtDecoder = require('../src/controllers/messageTypes/jwt/jsonwebtoken/decode');    
+
+describe('When signing an x509 certificate', function () {
+
```
Comment (lyzhang27) [inline test/x509-test.js]: This is an extra line.

File: test/x509-test.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    var token = basicIdToken.toJWT(priv, {algorithm: 'RS256'});
+    
+    it('should work with the corresponding pub key 1', function () {
+      var verified = jwtDecoder.verifyJwtSign(token, pub, basicIdToken,  {"clockTimestamp" : clockTimestamp, jwtid: 'jwtid', algorithm: 'HS256'}, "base64");
+      assert.equal(verified, true);
+    });
+});
+
+describe('When signing an x509 certificate', function () {
+
```
Comment (lyzhang27) [inline test/x509-test.js]: Extra line?

File: test/x509-test.js [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+  var priv = fs.readFileSync("/Users/anjuthomas/Documents/example_key.pem");
```
Comment (lyzhang27) [inline test/x509-test.js]: We should not be checking in this local directory path

File: test/x509-test.js [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+  var basicIdToken = new BasicIdToken('issuer','subject', clockTimestamp, "jwtid");
+  basicIdToken.addNonStandardClaims({"foo": 'bar', "aud" : "audience"});
+  basicIdToken.setNoneAlgorithm(true);
+  var token = basicIdToken.toJWT(priv, {algorithm: 'RS256'});
+  
+  it('should work with the corresponding pub key 2', function () {
+    var verified = jwtDecoder.verifyJwtSign(token, pub, basicIdToken,  {"clockTimestamp" : clockTimestamp, jwtid: 'jwtid', algorithm: 'HS256'}, "base64");
+    assert.equal(verified, true);
+  });
+
```
Comment (lyzhang27) [inline test/x509-test.js]: Extra line

File: test/x509-test.js [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+  it('should not work with wrong pub key 1', function () {
+    var pub = fs.readFileSync("/Users/anjuthomas/Documents/rolandPub.pem");      
+    var verified = jwtDecoder.verifyJwtSign(token, pub, basicIdToken,  {"clockTimestamp" : clockTimestamp, jwtid: 'jwtid', algorithm: 'HS256'}, "base64");
+    assert.equal(verified, false);
+  });
+});
+
+describe('When signing an x509 certificate', function () {
+    var priv = fs.readFileSync("/Users/anjuthomas/Documents/example_key.pem");
```
Comment (lyzhang27) [inline test/x509-test.js]: We should not be checking in this absolute local path.

File: test/x509-test.js [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+  var token = basicIdToken.toJWT(priv, {algorithm: 'RS256'});
+  
+  it('should work with the corresponding pub key 2', function () {
+    var verified = jwtDecoder.verifyJwtSign(token, pub, basicIdToken,  {"clockTimestamp" : clockTimestamp, jwtid: 'jwtid', algorithm: 'HS256'}, "base64");
+    assert.equal(verified, true);
+  });
+    
+
+  it('should not work with wrong pub key 1', function () {
+    var pub = fs.readFileSync("/Users/anjuthomas/Documents/rolandPub.pem");
```
Comment (lyzhang27) [inline test/x509-test.js]: We should not be checking in this absolute local path.

File: test/x509-test.js [hunk 14]
```diff
@@ -0,0 +1,66 @@
+var fs = require('fs');
+var BasicIdToken = require('../src/models/tokenProfiles/basicIdToken');
+var assert = require('chai').assert;  
+var XMLHttpRequest = require("xmlhttprequest").XMLHttpRequest;
+var jwtDecoder = require('../src/controllers/messageTypes/jwt/jsonwebtoken/decode');    
+
+describe('When signing an x509 certificate', function () {
+  
+    var priv = fs.readFileSync("/Users/anjuthomas/Documents/rolandPriv.pem");
```
Comment (lyzhang27) [inline test/x509-test.js]: We should not be checking in this absolute local path.

File: test/x509-test.js [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+    
+    it('should work with a JWK input', function () {
+      var verified = jwtDecoder.verifyJwtSignature(token, pub, basicIdToken,  {"clockTimestamp" : clockTimestamp, jwtid: 'jwtid', algorithm: 'HS256'}, JSON.stringify(jwk), "base64");
+      assert.equal(verified, true);
+    });
+  });
+
+
```
Comment (lyzhang27) [inline test/x509-test.js]: There are a couple of extra lines here.

**Review by lyzhang27 — CHANGES_REQUESTED** [review]

File: src/controllers/messageTypes/jwt/jsonwebtoken/decode.js (hunk 4 above)
Comment (anjuthomas) [reply src/controllers/messageTypes/jwt/jsonwebtoken/decode.js]: Done

File: src/controllers/messageTypes/jwt/jsonwebtoken/decode.js (hunk 1 above)
Comment (anjuthomas) [reply src/controllers/messageTypes/jwt/jsonwebtoken/decode.js]: Done

File: src/controllers/messageTypes/jwt/jsonwebtoken/decode.js (hunk 2 above)
Comment (anjuthomas) [reply src/controllers/messageTypes/jwt/jsonwebtoken/decode.js]: Done

File: src/controllers/messageTypes/jwt/jsonwebtoken/decode.js (hunk 5 above)
Comment (anjuthomas) [reply src/controllers/messageTypes/jwt/jsonwebtoken/decode.js]: Done

File: src/controllers/messageTypes/jwt/jsonwebtoken/decode.js (hunk 3 above)
Comment (anjuthomas) [reply src/controllers/messageTypes/jwt/jsonwebtoken/decode.js]: Done

File: src/controllers/messageTypes/jwt/jsonwebtoken/decode.js (hunk 6 above)
Comment (anjuthomas) [reply src/controllers/messageTypes/jwt/jsonwebtoken/decode.js]: Removed

File: src/controllers/messageTypes/jwt/jsonwebtoken/decode.js (hunk 7 above)
Comment (anjuthomas) [reply src/controllers/messageTypes/jwt/jsonwebtoken/decode.js]: Done

File: test/x509-test.js (hunk 8 above)
Comment (anjuthomas) [reply test/x509-test.js]: Done

File: test/x509-test.js (hunk 9 above)
Comment (anjuthomas) [reply test/x509-test.js]: Done

File: test/x509-test.js (hunk 15 above)
Comment (anjuthomas) [reply test/x509-test.js]: Done

File: test/x509-test.js (hunk 12 above)
Comment (anjuthomas) [reply test/x509-test.js]: Done

File: test/x509-test.js (hunk 13 above)
Comment (anjuthomas) [reply test/x509-test.js]: Done

File: test/x509-test.js (hunk 10 above)
Comment (anjuthomas) [reply test/x509-test.js]: Done

File: test/x509-test.js (hunk 11 above)
Comment (anjuthomas) [reply test/x509-test.js]: Done

File: test/x509-test.js (hunk 14 above)
Comment (anjuthomas) [reply test/x509-test.js]: Done

**Review by lyzhang27 — APPROVED** [review]

**Pull request merged**